### PR TITLE
feat(timetypes): add duration type

### PR DESCRIPTION
**Pull request by NikolaLohinski**
Support Go's duration type as a TF type.

### CODE REVIEW

File: .changes/unreleased/FEATURES-20240525-222251.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+kind: FEATURES
+body: Support go time duration type
```
Comment (bflad) [inline .changes/unreleased/FEATURES-20240525-222251.yaml:2]:
```suggestion
body: Support Go standard library time duration type (`GoDurationType` and `GoDuration`)
```

File: timetypes/diagnostics.go [hunk 2]
```diff
@@ -15,3 +15,14 @@ func rfc3339InvalidStringDiagnostic(value string, err error) diag.Diagnostic {
 			"Error: "+err.Error(),
 	)
 }
+
+// durationInvalidStringDiagnostic returns an error diagnostic intended to report
+// when a string is not a time duration.
+func durationInvalidStringDiagnostic(value string, err error) diag.Diagnostic {
+	return diag.NewErrorDiagnostic(
+		"Invalid time duration String Value",
```
Comment (bflad) [inline timetypes/diagnostics.go:23]: Nit: Many Terraform diagnostics are written in title case

```suggestion
		"Invalid Time Duration String Value",
```

File: timetypes/diagnostics.go [hunk 3]
```diff
@@ -15,3 +15,14 @@ func rfc3339InvalidStringDiagnostic(value string, err error) diag.Diagnostic {
...
+		"A string value was provided that is not valid time duration string format.\n\n"+
```
Comment (bflad) [inline timetypes/diagnostics.go:24]: When receiving an invalid value error diagnostic, practitioners will likely want to know what a valid format is so they can resolve the error. The `time` package errors can be fairly opaque about what is valid, such as:

```
time: invalid duration "!"
time: missing unit in duration "10"
time: unknown unit "d" in duration "10d"
```

https://pkg.go.dev/time#ParseDuration mentions what is valid as:

> A duration string is a possibly signed sequence of decimal numbers, each with optional fraction and a unit suffix, such as "300ms", "-1.5h" or "2h45m". Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".

It does not need to be verbatim, but we should include this information in the diagnostic, such as:

```suggestion
		"A string value was provided that is not valid time duration string format. "+
			"A duration string is a sequence of numbers, each with optional fraction and a unit suffix, such as "300ms", "-1.5h" or "2h45m". "+
			"Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".\n\n"+
```

File: timetypes/duration_type.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+var (
+	_ basetypes.StringTypable = (*DurationType)(nil)
+)
+
+// DurationType is an attribute type that represents a valid Go time duration string.
+// See https://pkg.go.dev/time#ParseDuration for more details
+type DurationType struct {
```
Comment (bflad) [inline timetypes/duration_type.go:22]: Given that the Go standard library `time` package is just one potential time duration implementation, we should explicitly define the type and value type to differentiate it from other potential duration implementations in the future. For example, [ISO 8601 implements its own duration format](https://en.wikipedia.org/wiki/ISO_8601#Durations).

```suggestion
type GoDurationType struct {
```

This change will need to be applied elsewhere, such as method signatures and file naming.

File: timetypes/duration_type_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+func TestDurationTypeValueFromTerraform(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		in          tftypes.Value
+		expectation attr.Value
+		expectedErr string
+	}{
+		"true": {
```
Comment (bflad) [inline timetypes/duration_type_test.go:24]: Nit: not sure what "true" is here, but I think this unit test is trying to test a valid value

```suggestion
		"value": {
```

File: timetypes/duration_type_test.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+			got, err := timetypes.DurationType{}.ValueFromTerraform(ctx, testCase.in)
+			if err != nil {
+				if testCase.expectedErr == "" {
+					t.Fatalf("Unexpected error: %s", err)
+				}
+				if testCase.expectedErr != err.Error() {
+					t.Fatalf("Expected error to be %q, got %q", testCase.expectedErr, err.Error())
+				}
+				return
```
Comment (bflad) [inline timetypes/duration_type_test.go:55]: Nit: Given use of `(*testing.T).Fatalf()`, which will immediately stop the test, `return` is not necessary

```suggestion
```

File: timetypes/duration_value.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+var (
+	_ basetypes.StringValuable       = (*Duration)(nil)
+	_ xattr.ValidateableAttribute    = (*Duration)(nil)
+	_ function.ValidateableParameter = (*Duration)(nil)
+)
+
+// Duration represents a valid Go time duration string.
+// See https://pkg.go.dev/time#ParseDuration for more details
+type Duration struct {
```
Comment (bflad) [inline timetypes/duration_value.go:26]: Similarly with the value type

```suggestion
type GoDuration struct {
```

Along with changes elsewhere.

**Review by bflad — CHANGES_REQUESTED** [review]
Hello, @NikolaLohinski 👋 Thank you for submitting this. Please note that in the future, it is generally preferable for many HashiCorp-owned projects to create a feature request issue and ensure the maintainers approve of the design before beginning implementation work. This can potentially save you a lot of code writing that potentially might not be accepted or might need to be rewritten. Luckily in this case, supporting the Go standard library `time` package `Duration` type does fall under an acceptable implementation for this Go module's intended design and scope as well as the fact that many framework custom types typically follow a consistent implementation.

Overall this is looking good, but I think we'll want to be careful about the type naming including "Go" or something similar to differentiate it from other potential time duration formats, such as ISO 8601. Once the changes are in, this will likely be ready for a final review.

Thanks again for this contribution!

File: timetypes/diagnostics.go (hunk 3 above)
Comment (NikolaLohinski) [reply timetypes/diagnostics.go]: Done :heavy_check_mark:

File: timetypes/duration_type.go (hunk 4 above)
Comment (NikolaLohinski) [reply timetypes/duration_type.go]: Done :heavy_check_mark:

File: timetypes/duration_value.go (hunk 7 above)
Comment (NikolaLohinski) [reply timetypes/duration_value.go]: Done :heavy_check_mark:

**Comment by NikolaLohinski** [comment]
> Hello, @NikolaLohinski 👋 Thank you for submitting this. Please note that in the future, it is generally preferable for many HashiCorp-owned projects to create a feature request issue and ensure the maintainers approve of the design before beginning implementation work. This can potentially save you a lot of code writing that potentially might not be accepted or might need to be rewritten. Luckily in this case, supporting the Go standard library `time` package `Duration` type does fall under an acceptable implementation for this Go module's intended design and scope as well as the fact that many framework custom types typically follow a consistent implementation.
> 
> Overall this is looking good, but I think we'll want to be careful about the type naming including "Go" or something similar to differentiate it from other potential time duration formats, such as ISO 8601. Once the changes are in, this will likely be ready for a final review.
> 
> Thanks again for this contribution!

Hi @bflad and thanks for taking the time to review my PR.

I am very well aware of the contribution guidelines, but this seemed so straightforward and logical with regards to the design that I just went ahead and built it in order to leverage it in https://github.com/NikolaLohinski/terraform-provider-freebox and avoid some back and forth interactions with an issue.

I've taken your comments into account and renamed things where it made sense (almost everywher :rofl: ).

Let me know if anything else is needed :pray: 

Cheers

File: timetypes/go_duration_value_example_test.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework-timetypes/timetypes"
+)
+
+type DurationResourceModel struct {
+	Duration timetypes.GoDuration `tfsdk:"duration"`
+}
+
+func ExampleGoDuration_ValueDuration() {
```
Comment (bflad) [inline timetypes/go_duration_value_example_test.go:16]: Looks like golangci-lint is still showing one report:

```
  Error: timetypes/go_duration_value_example_test.go:16:1: tests: ExampleGoDuration_ValueDuration refers to unknown field or method: GoDuration.ValueDuration (govet)
  func ExampleGoDuration_ValueDuration() {
  ^
```

```suggestion
func ExampleGoDuration_ValueGoDuration() {
```

Comment (NikolaLohinski) [reply timetypes/go_duration_value_example_test.go]: Yeah I caught that a few minutes ago and fixed it as well. Thanks

**Review by bflad — APPROVED** [review]
Looks good to me, thank you, @NikolaLohinski! 🚀 Will merge and cut a release shortly.

**Comment by NikolaLohinski** [comment]
Awesome, thanks a ton :bow:

**Pull request merged**